### PR TITLE
The firewaller is refactored to push ingress changes to the offering model instead of polling it

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

TL;DR: the firewaller pushes ingress changes to the offering model instead of that model polling the consuming model. This PR needs a follow up which will have the offering model react to the published changes.

There's also lot of code shuffling which accounts for a lot of the line count. The remotefirewaller facade is removed and the ingress address watcher moved to the firewaller facade. The crossmodelrelations facade also gains a new publish ingress changes method.
The firewaller now only needs to start one watcher to monitor address changes, not two. The new code copied across from the delete remote firewaller facade is unit tested whereas the other firewaller tests are jujuconnsuite, so accommodations are made for that.

## QA steps

Deploy a mediawiki-mysql cross model relation and check logs to see that ingress change is published to mysql model.

### CODE REVIEW

File: api/firewaller/firewaller.go [hunk 1]
```diff
@@ -130,3 +130,25 @@ func (st *State) Relation(tag names.RelationTag) (*Relation, error) {
 		life: life,
 	}, nil
 }
+
+// WatchIngressAddressesForRelation returns a watcher that notifies when address,
```
Comment (axw) [inline api/firewaller/firewaller.go]: addresses?

File: apiserver/allfacades.go [hunk 2]
```diff
@@ -135,7 +134,7 @@ func AllFacades() *facade.Registry {
 	reg("Controller", 3, controller.NewControllerAPI)
 	reg("Deployer", 1, deployer.NewDeployerAPI)
 	reg("DiskManager", 2, diskmanager.NewDiskManagerAPI)
-	reg("Firewaller", 3, firewaller.NewFirewallerAPI)
+	reg("Firewaller", 3, firewaller.NewStateFirewallerAPI)
```
Comment (axw) [inline apiserver/allfacades.go]: bump the version? can we claim support for this new method? maybe it needs to be added to the facade only when the feature flag is set?

File: apiserver/firewaller/firewaller.go [hunk 3]
```diff
@@ -347,3 +359,120 @@ func (f *FirewallerAPI) getMachine(canAccess common.AuthFunc, tag names.MachineT
 	// machine.
 	return entity.(*state.Machine), nil
 }
+
+// WatchIngressAddressesForRelation creates a watcher that notifies when address, from which
```
Comment (axw) [inline apiserver/firewaller/firewaller.go]: addresses?

File: apiserver/allfacades.go (hunk 2 above)
Comment (wallyworld) [inline apiserver/allfacades.go]: I need to bump the version for sure. Created v4 and only register if flag is set.

File: api/firewaller/firewaller.go (hunk 1 above)
Comment (wallyworld) [inline api/firewaller/firewaller.go]: yep

File: apiserver/firewaller/firewaller.go (hunk 3 above)
Comment (wallyworld) [inline apiserver/firewaller/firewaller.go]: yep

File: worker/firewaller/firewaller.go [hunk 4]
```diff
@@ -141,15 +144,15 @@ func NewFirewaller(cfg Config) (worker.Worker, error) {
...
+		newRemoteFirewallerAPIFunc: cfg.NewCrossModelFacadeFunc,
 		modelUUID:                  cfg.ModelUUID,
 		machineds:                  make(map[names.MachineTag]*machineData),
 		unitsChange:                make(chan *unitsChange),
 		unitds:                     make(map[names.UnitTag]*unitData),
 		applicationids:             make(map[names.ApplicationTag]*applicationData),
 		exposedChange:              make(chan *exposedChange),
 		relationIngress:            make(map[names.RelationTag]*remoteRelationData),
-		remoteRelationsChange:      make(chan *remoteRelationChange),
+		remoteRelationsChange:      make(chan *remoteRelationNetworkChange),
```
Comment (wallyworld) [inline worker/firewaller/firewaller.go]: ok

File: apiserver/common/crossmodel/crossmodel.go [hunk 5]
```diff
@@ -182,3 +182,45 @@ func RelationUnitSettings(backend Backend, ru params.RelationUnit) (params.Setti
...
+	// Look up the application on the remote side of this relation
+	// ie from the model which published this change.
+	applicationTag, err := getRemoteEntityTag(backend, change.ApplicationId)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	logger.Debugf("application tag for remote id %+v is %v", change.ApplicationId, applicationTag)
+
+	// TODO(wallyworld) - deal with remote application being removed
+	if applicationTag == nil {
```
Comment (wallyworld) [inline apiserver/common/crossmodel/crossmodel.go]: fixed, was some legacy code there

File: apiserver/common/crossmodel/crossmodel.go [hunk 6]
```diff
@@ -182,3 +182,45 @@ func RelationUnitSettings(backend Backend, ru params.RelationUnit) (params.Setti
...
+
+// PublishIngressNetworkChange saves the specified ingress networks for a relation.
+func PublishIngressNetworkChange(backend Backend, change params.IngressNetworksChangeEvent) error {
+	logger.Debugf("publish into model %v network change: %+v", backend.ModelUUID(), change)
+
+	relationTag, err := getRemoteEntityTag(backend, change.RelationId)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Debugf("not found relation tag %+v in model %v, exit early", change.RelationId, backend.ModelUUID())
+			return nil
```
Comment (wallyworld) [inline apiserver/common/crossmodel/crossmodel.go]: ok

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$
